flex_model/distributed/mappings.py

```diff
@@ -237,30 +238,298 @@ def scatter_data_parallel(tensor: Tensor, dim: int = 0) -> Tensor:
+
+        # Add gathered tbufs.
+        for recv_tbuf, recv_r in zip(all_recv_tbufs, all_recv_ranks):
+            dtype = recv_tbuf.dtype
+            meta = all_metadata_groups[recv_r][dtype]
+
+            buf_rank = meta["buffer_rank"]
+            buf_dtype = meta["buffer_dtype"]
+            assert buf_dtype == dtype, (
+                f"Dtype mismatch: {buf_dtype} and {dtype}")
+            assert buf_rank == recv_r, (
+                f"Rank mismatch: {buf_rank} and {recv_r}")
+
+            _reshard_tbuf(meta, recv_tbuf)
+
+    return output_tensor_dict
+
+
+def batch_isend_irecv_pipeline_parallel(
+    recv_tensors: List[Tensor],
+    recv_from_ranks: List[int],
+    send_tensors: List[Tensor],
+    send_to_ranks: List[int],
+) -> None:
+    """Run batched peer-to-peer communications.
+
+    :param List[Tensor] recv_tensors: Tensors to receive.
+    :param List[int] recv_from_ranks: Ranks to receive from.
+    :param List[Tensor] send_tensors: Tensors to send.
+    :param List[int] send_to_ranks: Ranks to send to.
+    """
+    world_size = dist.get_activation_pipeline_parallel_world_size()
+    rank = dist.get_activation_pipeline_parallel_rank()
+    group = dist.get_activation_pipeline_parallel_group()
+
+    assert len(recv_tensors) == len(recv_from_ranks), (
+        f"Mistmatch in recv tensors({len(recv_tensors)}) and "
+        f"recv ranks({len(recv_from_ranks)})"
+    )
+    assert len(send_tensors) == len(send_to_ranks), (
+        f"Mistmatch in send tensors({len(send_tensors)}) and "
+        f"send ranks({len(send_to_ranks)})"
+    )
+
+    p2p_ops = []
+    for recv_t, recv_r in zip(recv_tensors, recv_from_ranks):
+        op = torch.distributed.P2POp(
+            torch.distributed.irecv,
+            recv_t,
+            peer=recv_r,
+            group=group,
+        )
+        p2p_ops.append(op)
+
+        logger.debug(f"Rank{rank}: P2POp (irecv) [{rank}] <- [{recv_r}]")
+
+    for send_t, send_r in zip(send_tensors, send_to_ranks):
+        op = torch.distributed.P2POp(
+            torch.distributed.isend,
+            send_t,
+            peer=send_r,
+            group=group,
+        )
+        p2p_ops.append(op)
+
+        logger.debug(f"Rank{rank}: P2POp (isend) [{rank}] -> [{send_r}]")
+
+    if len(p2p_ops) == 0:
+        return
 
-    Takes a pickle-able collection of python objects and tensors and sends
-    them to rank0 in the pipeline parallel group.
+    logger.debug(f"Rank{rank}: Launching P2POps")
 
-    :param Any objects: Some python object that can be pickled. May contain tensors.
+    reqs = torch.distributed.batch_isend_irecv(p2p_ops)
+    for req in reqs:
+        req.wait()
+
+    _msg = lambda t_list: ", ".join([f"({t.numel()}, {t.dtype})" for t in t_list])
+    logger.debug(f"Rank{rank}: Received buffers - [{_msg(recv_tensors)}]")
+    logger.debug(f"Rank{rank}: Sent buffers - [{_msg(send_tensors)}]")
+
+    # TODO: Remove after verification that no race cond. occurs.
```

Comment:
Should this be captured as an issue or ticket?

Reply:
This was an unresolved TODO in the `Megatron-LM` codebase which used the `torch.distributed.batch_isend_irecv()` function. Just keeping it here as a warning for now.